### PR TITLE
cmake: make library discoverable via find_package()

### PR DESCRIPTION
**Pull request by hgkjshegfskef**
Library can be discovered through find_package(CONFIG) mode. Library
produces Trace::trace IMPORTED target to be linked against. The target
is discoverable both in installed state, and when the project is
included as subdirectory.

Additionally, a few minor cleanups are made:
- no need to look for C compiler, there are no C files
- no need to print information about C compiler (no need to print
  information about C++ compiler either, since this is done by CMake
  automatically, but since it exists already, it is probably for a
  reason)
- library version is specified (required for target versioning)
- enable_testing() is wrapped inside BUILD_TESTING, which is a standard
  CMake variable, and set to ON by default to preserve existing behavior

After installation (tested) or inclusion as subdirectory (not tested),
expected usage is:

    find_package(Trace CONFIG REQUIRED)
    target_link_libraries(foobar PRIVATE Trace::trace)

Library can be included via:

    #include <trace/trace.hpp>

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,13 +9,12 @@ option( TRACE_TEST_COVERAGE "Build trace tests with coverage flags" OFF )
 
 
 ######### Project Name ####################################
-project( trace )
+project( Trace LANGUAGES CXX VERSION 0.1 )
```
Comment (GPMueller) [inline CMakeLists.txt:13]: Any reason to change the name to uppercase `Trace`?

File: CMakeLists.txt [hunk 2]
```diff
@@ -63,51 +62,102 @@ set( TRACE_SOURCE_FILES
...
+include( GNUInstallDirs )
+
 set_target_properties( ${LIBRARY_NAME} PROPERTIES
     CXX_STANDARD          17
     CXX_STANDARD_REQUIRED ON
     CXX_EXTENSIONS        OFF
     COMPILE_FLAGS         "${TRACE_COMPILE_FLAGS} ${TRACE_FLAGS_COVERAGE}"
-    LINK_FLAGS            "${TRACE_LINK_FLAGS}    ${TRACE_FLAGS_COVERAGE}" )
+    LINK_FLAGS            "${TRACE_LINK_FLAGS}    ${TRACE_FLAGS_COVERAGE}"
+    VERSION               "${${PROJECT_NAME}_VERSION}"
```
Comment (GPMueller) [inline CMakeLists.txt:74]: I believe `PROJECT_NAME` is set by the topmost `project(...)` command, so it would be safer to use `LIBRARY_NAME` here and in the following

File: CMakeLists.txt [hunk 3]
```diff
@@ -63,51 +62,102 @@ set( TRACE_SOURCE_FILES
...
+                     "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake"
+         DESTINATION "${${PROJECT_NAME}_CMAKE_FILES_DESTINATION}" )
+
+export( EXPORT ${PROJECT_NAME}Targets
+        FILE   "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Targets.cmake" )
+###########################################################
+
+
+######### Tests ###########################################
+option( BUILD_TESTING ON )
```
Comment (GPMueller) [inline CMakeLists.txt]: Please place `option`s near the top of the CMakeLists.txt, so they are obvious when reading this file.

Something about your changes (I cannot currently see the reason) is also keeping the tests from being built on the CI, which is therefore currently failing.

**Review by GPMueller — CHANGES_REQUESTED** [review]
Thanks for your PR!

As I merged your PR #2, please remove your second commit (to remove the semicolons) from this PR.

If you can get the CI running again and apply the small changes I mentioned, I'd be happy to merge your PR.

File: CMakeLists.txt (hunk 1 above)
Comment (hgkjshegfskef) [reply CMakeLists.txt:13]: Yes, this is to use `${PROJECT_NAME}` without directly using "Trace" to refer to the namespace of the target, as in, `Trace::trace` (usually, target's namespace is capitilized in CMake code, but it is not a firmly established practice), so that perhaps somebody can look at the code and practically copy it verbatim, as the procedure of creating and exporting targets and configs is pretty much boilerplate. I considered changing project name to be acceptable, since target name `trace` is still lowercase.

File: CMakeLists.txt (hunk 2 above)
Comment (hgkjshegfskef) [reply CMakeLists.txt:74]: There is `PROJECT_NAME` and `CMAKE_PROJECT_NAME`. The former always refers to the innermost project, the latter always refers to the outermost project. That is, if somebody adds this library as subdirectory, CMAKE_PROJECT_NAME will refer to their project, and PROJECT_NAME will refer to this library.

File: CMakeLists.txt (hunk 3 above)
Comment (hgkjshegfskef) [reply CMakeLists.txt]: Right. I remember now: I if'ed all the tests because I couldn't compile `catch` due to https://github.com/catchorg/Catch2/issues/2178 . You should probably update the header.

**Comment by hgkjshegfskef** [comment]
Forgot to write, this is ready for another review.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -18,13 +18,21 @@ jobs:
     steps:
     - uses: actions/checkout@v2
 
+    - name: 📚 Install necessary packages
+      shell: bash
+      run:  |
+            if [ "$RUNNER_OS" == "Linux" ]; then
```
Comment (GPMueller) [inline .github/workflows/ci.yml]: Instead of using bash's if, please use the GH Actions way of doing it with `if: matrix.os == 'ubuntu-latest'`

**Review by GPMueller — CHANGES_REQUESTED** [review]
Except the `if` in the CI script, everything looks very good 👍
We can ignore the minor coverage change perceived by codecov.

**Comment by hgkjshegfskef** [comment]
New changes, fixed up to last commit:
- styled if check as requested
- removed if check on coverage job, since I realized it only runs on ubuntu
- added bash shell in 'create build dir' step for consistency with all other steps

Ready for another review.

**Comment by hgkjshegfskef** [comment]
My bad, working-directory broke.

**Comment by hgkjshegfskef** [comment]
Now the CI is not triggered at all. I don't get it.

**Comment by hgkjshegfskef** [comment]
CI fixed.

Changes:
- make bash shell the default shell for all jobs
- use single quotes around ${{runner.workspace}} to prevent bash from
  treating the '\' windows path separators as escapes (thanks Donovan
  Baarda on SO)
- "steps:" indentation matches github actions docs' examples

Ready for review.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -16,64 +24,66 @@ jobs:
...
-    - uses: actions/checkout@v2
+      - uses: actions/checkout@v2
 
-    - name: 📁 Create build folder
-      run: cmake -E make_directory ${{runner.workspace}}/build
+      - name: 📚 Install necessary packages
+        if: matrix.os == 'ubuntu-latest'
+        run: |
+          sudo apt update
+          sudo apt install -y cmake
```
Comment (GPMueller) [inline .github/workflows/ci.yml:33]: why did you add an installation for cmake? I don't believe it's necessary

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -16,64 +24,66 @@ jobs:
...
-      uses: codecov/codecov-action@v1
-      with:
-        files: ${{runner.workspace}}/build/coverage.info
\ No newline at end of file
+      - uses: actions/checkout@v2
+
+      - name: 📚 Install necessary packages
+        run: |
+          sudo apt update
+          sudo apt install -y cmake python3-pip lcov
```
Comment (GPMueller) [inline .github/workflows/ci.yml:62]: why did you add `python3-pip` to the installed packages? I don't think it's necessary

File: .github/workflows/ci.yml (hunk 5 above)
Comment (hgkjshegfskef) [reply .github/workflows/ci.yml:33]: Apparently, https://github.com/nektos/act is not configured to use github's runners by default, and I was using it to test CI locally, which was failing due to missing dependenecies. And apparently, github images ship with a ginormous database of preinstalled packages (13G). Therefore, you are correct, it is not necessary. However, I suggest to keep this in order to not have to have the github's giant image with all the packages stored and updated locally on disk, because local CI runs become a pain.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (hgkjshegfskef) [reply .github/workflows/ci.yml:62]: Same as the above.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (GPMueller) [reply .github/workflows/ci.yml:33]: hmm interesting consideration. I tend to be minimalist, i.e. to only install what I really need, but this is fine

File: CMakeLists.txt (hunk 3 above)
Comment (GPMueller) [reply CMakeLists.txt]: I created a corresponding issue #3

**Review by GPMueller — APPROVED** [review]
